### PR TITLE
Migrate AttributeValue

### PR DESCRIPTION
**Pull request by RanVaknin**
## Scope

This PR ports only the converter/marshalling layer of the v1 `DynamoDBMapper` to the v2 `AttributeValue` type (`software.amazon.awssdk.services.dynamodb.model.AttributeValue`). Every operation on `DynamoDBMapper` (load, save, query, scan, batch, transaction, tableAdmin) stays on v1 and is ported in later PRs.

**The PR does not compile** - The files that stay v1 (the two operation files, the paginated list classes that feed v1 results into v2 `toParameters`, and `DynamoDBMapperFieldModel`'s query condition methods) reference v1 types that no longer line up with the v2 converter surface, so those files are red. This keeps the diff scoped to the converter layer, and the subsequent operation PRs turn it green. 

This is intentional in order to make each PR review boundary conceptually separate and limits the diff. 

## Testing approach

- **Baseline on v1** - Run existing + new test against the v1 source before porting the types to v2
    - Existing converter tests all pass against the v1 source before any change
    - Added a new edge case suite (`StandardModelFactoriesEdgeCasesTest`) to widen coverage of the boundaries this migration touches (empty vs unset collections, binary roundtrips, boolean encodings, date formats)

- **Migrate the test bodies, not the assertions.** For each ported test we changed only the mechanical v1→v2 surface in the test body (`new AttributeValue().withS(x)` → `AttributeValue.builder().s(x).build()`, `getS()` → `s()`, and so on) The assertion logic stayed the same (aside from migrating the new types needed for assertion). A green run therefore means the v2 source produces the same result the v1 source did.
- **Running them despite the red module.** The `maven-compiler-plugin` `testIncludes` allowlist in `pom.xml` compiles only the converter tests. To execute them we temporarily stubbed the operation files (and their closure), compiled, and ran the suite: 287 passed, 0 failed. The stub was then reverted, restoring the red state.

#### 1. `set()`/`with()` → `builder().build()`
- v1 `AttributeValue` is mutable (`new AttributeValue().withS(x)`); 
- v2 is an immutable builder (`AttributeValue.builder().s(x).build()`). 

**Question** should we use the fast `createX()` factory methods added in #7039 here in this PR?
I decided to not port these here so we can measure performance against a baseline port vs performance improvement branch later.

#### 2. null vs empty — `has` guards

v2 `.l()`/`.m()`/`.ss()`/`.ns()`/`.bs()` return empty collections (not null) when unset. v2 added `hasL()`/`hasM()`/`hasSs()`/`hasNs()`/`hasBs()` to distinguish "explicitly empty" from "unset". Guards were added in the parent `typeCheck` methods (`LUnmarshaller`, `MUnmarshaller`, `SSUnmarshaller`, etc.), which run before `unmarshall` via `ConversionSchemas` dispatch. Pattern: `value.hasSs() ? value.ss() : null`.

#### 3. Boolean legacy `"1"/"0"` encoding

Old DynamoDB stored booleans as Number `"1"/"0"` before native `BOOL` existed. The reader accepts both and produces the same `Boolean`; the writer produces the number form.

`V2CompatibleBool.get`: `if (o.bool() != null) return o.bool() ? "1" : "0"; return o.n();`

Covered by `StandardModelFactoriesV2UnconvertTest` — `n("0")`/`n("1")`, `bool(false)`/`bool(true)`, and boolean set from `ns("1")`/`ns("0")`/`ns("0","1")`.

#### 3. SdkBytes boundary

v1 binary is `java.nio.ByteBuffer`; v2 binary is `software.amazon.awssdk.core.SdkBytes`. On write, `SdkBytes.fromByteBuffer(bb)` copies the buffer and leaves the caller's position untouched, matching v1.

On read, POJO fields typed `ByteBuffer` must stay mutable to match v1's `getB()`. `SdkBytes` which AttributeValue uses internally, is immutable and exposes its bytes only as a readonly buffer (via `asByteBuffer()`) so every `ByteBuffer` read site returns `ByteBuffer.wrap(sdkBytes.asByteArray())`. **The copy costs one O(n) pass per read, so this path is slower than v1**

Returning `asByteBuffer()` is arguably the better, more v2 idiomatic choice, but if a v1 customer tries to mutate a buffer between reads and writes, the immutable implementation would throw `ReadOnlyBufferException` at runtime. 

As part of future work I plan to add an opt-in feature (field annotation or mapper config flag) that returns the readonly zero copy buffer instead of the mutable copy, for callers who never mutate and want to beat v1's read performance. Same `ByteBuffer` field type, behavior selected by the flag.

**Question** - Should we pursue backwards compatibility? Or limit potential performance regression? Or add support for it via future opt in flag? (supporting both cases)

#### 5. `MapperDateUtils.java` (new file)

The date converters format and parse ISO-8601 strings through two methods on v1's `com.amazonaws.util.DateUtils`. A v2 module can't depend on that class, and v2's `software.amazon.awssdk.utils.DateUtils` isn't equivalent: it works in `java.time.Instant` rather than `Date`, and its formatter omits zero milliseconds (`...T00:00:00Z` vs v1's `...T00:00:00.000Z`), so switching to it would change the stored string form of dates and break round-trip compatibility with v1 data.

`MapperDateUtils` ports only the two methods the converters use, on v1's Joda-Time formatters, to keep that format identical. A partial port is enough since the converter layer touches nothing else in `DateUtils`.

#### 6. MapperExceptions.java (new file) 

Both marshallers wrapped errors with v1's `Throwables.failure`, which a v2 module can't depend on. Instead we ported the `failure` function helper into a new class `MapperExceptions.java`

**Breaking change:** checked exceptions now wrap in `SdkClientException` instead of v1's `AmazonClientException`. The breaking change is unavoidable, so users who are explicitly error handling the v1 exception would have to update their code beyond the promised "import swap". 

**Question** - Is the name for this new class acceptable? Should we port this method as a standalone and throw it in some other util package we have in the SDK?

---

```diff
DDB mapper v2 roadmap

** BASE PACKAGE SETUP **
+ 1. Source verbatim port ✅
+ 2. Test verbatim port ✅
+ 3. Namespace swap (main + datamodeling tests) ✅ 
+ 4. Namespace swap (remaining test packages) ✅

** PORTING OPERATIONS **
0. converters (AttributeValue seam)  <---- current PR
1. load()
2. save()
3. query() + scan()
4. deleteItem()
5. updateItem()
6. Batch operations
7. Transactions
8. Misc

** PERFORMANCE IMPROVEMENTS **
1. getTableModel caching
2. Wire "fast" createX AV factory methods to convertors
3. ByteBuffer → SdkBytes copy
4. Others

** DEPENDENCY MODERNIZATION **
1. EasyMock -> Mockito
2. Log4j 1.x -> 2.x
3. commons-logging -> SLF4J
```

### CODE REVIEW

File: services-custom/dynamodb-mapper/src/test/java/software/amazon/awssdk/mapper/dynamodb/StandardModelFactoriesEdgeCasesTest.java [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import software.amazon.awssdk.core.SdkBytes;
+import software.amazon.awssdk.services.dynamodb.model.AttributeValue;
+import software.amazon.awssdk.mapper.dynamodb.pojos.AutoKeyAndVal;
+import software.amazon.awssdk.mapper.dynamodb.pojos.TestClass;
+
+public class StandardModelFactoriesEdgeCasesTest {
```
Comment (dagnir) [inline services-custom/dynamodb-mapper/src/test/java/software/amazon/awssdk/mapper/dynamodb/StandardModelFactoriesEdgeCasesTest.java:37]: Can you add a javadoc for this class? What specific edge cases are we testing? why?

File: services-custom/dynamodb-mapper/src/main/java/software/amazon/awssdk/mapper/dynamodb/MapperDateUtils.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+import software.amazon.awssdk.annotations.SdkInternalApi;
+
+/**
+ * A thin port of only the ISO-8601 parse/format methods of the v1 {@code com.amazonaws.util.DateUtils}
+ * that the mapper's converter layer uses, kept on the same Joda-Time formatters to preserve byte-for-byte
+ * compatibility with dates written by the v1 mapper.
+ */
+@SdkInternalApi
+public final class MapperDateUtils {
```
Comment (dagnir) [inline services-custom/dynamodb-mapper/src/main/java/software/amazon/awssdk/mapper/dynamodb/MapperDateUtils.java:35]: are there tests for this?

File: services-custom/dynamodb-mapper/src/main/java/software/amazon/awssdk/mapper/dynamodb/MapperDateUtils.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+import java.util.Arrays;
+import java.util.Date;
+import java.util.List;
+
+import org.joda.time.DateTimeZone;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+import org.joda.time.format.ISODateTimeFormat;
+import org.joda.time.tz.FixedDateTimeZone;
```
Comment (dagnir) [inline services-custom/dynamodb-mapper/src/main/java/software/amazon/awssdk/mapper/dynamodb/MapperDateUtils.java:25]: Do we need to keep using Joda time for this internal class?

File: services-custom/dynamodb-mapper/src/main/java/software/amazon/awssdk/mapper/dynamodb/StandardModelFactories.java [hunk 4]
```diff
@@ -406,11 +406,24 @@ public DynamoDBTypeConverter<AttributeValue,Collection<T>> newConverter(Converti
...
             @Override
             public List<ByteBuffer> get(AttributeValue value) {
-                return value.getBS();
+                if (!value.hasBs()) {
+                    return null;
+                }
+                final List<ByteBuffer> result = new ArrayList<ByteBuffer>(value.bs().size());
+                for (SdkBytes sb : value.bs()) {
+                    // Writable copy for v1 parity; SdkBytes.asByteBuffer() would be read-only.
+                    result.add(ByteBuffer.wrap(sb.asByteArray()));
```
Comment (dagnir) [inline services-custom/dynamodb-mapper/src/main/java/software/amazon/awssdk/mapper/dynamodb/StandardModelFactories.java]: Creating a writable `ByteBuffer` from `SdkByte`s seems to occur often enough that it might make sense to create a util method

File: services-custom/dynamodb-mapper/src/main/java/software/amazon/awssdk/mapper/dynamodb/MapperDateUtils.java (hunk 3 above)
Comment (RanVaknin) [reply services-custom/dynamodb-mapper/src/main/java/software/amazon/awssdk/mapper/dynamodb/MapperDateUtils.java:25]: We don't need to keep using Joda here, but the converter layer itself uses Joda to do date parsing logic across multiple classes. Removing it from this specific utils class is not going to be enough to remove Joda as a dependency for this module.

I also looked into completely removing Joda from the implementation of those classes, but that will be a breaking change at runtime. If a user has a DDB model class with a field of type `org.joda.time.DateTime`, the mapper will try to match the right converter for that type, but will end up throwing:

https://github.com/aws/aws-sdk-java-v2/blob/56be1cd584410433d0a7d89af19b52c31e046c29/services-custom/dynamodb-mapper/src/main/java/software/amazon/awssdk/mapper/dynamodb/StandardTypeConverters.java#L513-L515

File: services-custom/dynamodb-mapper/src/main/java/software/amazon/awssdk/mapper/dynamodb/StandardModelFactories.java (hunk 4 above)
Comment (RanVaknin) [reply services-custom/dynamodb-mapper/src/main/java/software/amazon/awssdk/mapper/dynamodb/StandardModelFactories.java]: done, this will need a surface API review though for the new method.

File: services-custom/dynamodb-mapper/src/main/java/software/amazon/awssdk/mapper/dynamodb/MapperDateUtils.java (hunk 2 above)
Comment (RanVaknin) [reply services-custom/dynamodb-mapper/src/main/java/software/amazon/awssdk/mapper/dynamodb/MapperDateUtils.java:35]: good catch. I forgot to port the v1 tests for this class. I now ported them over from v1 (the ones that are relevant for these two ported methods)

File: services-custom/dynamodb-mapper/src/test/java/software/amazon/awssdk/mapper/dynamodb/StandardModelFactoriesEdgeCasesTest.java (hunk 1 above)
Comment (RanVaknin) [reply services-custom/dynamodb-mapper/src/test/java/software/amazon/awssdk/mapper/dynamodb/StandardModelFactoriesEdgeCasesTest.java:37]: >Can you add a javadoc for this class?

Added javadoc

>What specific edge cases are we testing?

I ported the existing v1 tests, but they only cover the happy path conversion for each type, not so much edge cases. The added coverage is unicode and emoji strings, empty strings, precision and limits of number types, JSON payloads, that Boolean schema difference between v1 and v2, and binary buffer which is new behavior. 

>why?

Mainly because Im paranoid about silent regressions across the major version boundary. This paid off in the past, for example the emoji test case revealed an issue with previous ddb work: https://github.com/aws/aws-sdk-java-v2/pull/6507

**Comment by dagnir** [comment]
> Question - Should we pursue backwards compatibility? Or limit potential performance regression? Or add support for it via future opt in flag? (supporting both cases)

We should favor backwads compat for `ByteBuffer`, so they should be mutable; if they have `SdkBytes` then use they should remain immutable.

> Question - Is the name for this new class acceptable? Should we port this method as a standalone and throw it in some other util package we have in the SDK?

This is fine, but we should keep it in the `mapper` package instead of leaking it elsewhere.

File: services-custom/dynamodb-mapper/src/main/java/software/amazon/awssdk/mapper/dynamodb/MapperDateUtils.java (hunk 3 above)
Comment (dagnir) [reply services-custom/dynamodb-mapper/src/main/java/software/amazon/awssdk/mapper/dynamodb/MapperDateUtils.java:25]: discussed this offline, but we should see if we can have just an `optional` dependency on `jodatime` so that it's not on the mapper dependency graph by default and customers using joda types in their POJOs can opt in by including the joda time dependency in their own POM.

Comment (RanVaknin) [reply services-custom/dynamodb-mapper/src/main/java/software/amazon/awssdk/mapper/dynamodb/MapperDateUtils.java:25]: I've looked further into it. Adding joda as an optional dependency wont work here as is. even if a customer has no joda typed field on their data class, the first converter resolution eagerly loads the `Scalar` enum, and several of its JDK type constants (`CALENDAR`, `DATE`, `LONG`) bake a `DateTime.class` literal into their converter maps, plus theres a dedicated Joda `DATE_TIME` enum value:  https://github.com/aws/aws-sdk-java-v2/blob/56be1cd584410433d0a7d89af19b52c31e046c29/services-custom/dynamodb-mapper/src/main/java/software/amazon/awssdk/mapper/dynamodb/StandardTypeConverters.java#L169 
 
The only way I can see to make it work is a runtime check that inspects whether a field is a joda type before deciding which converters to use, likely by name or reflection.
That adds a branch on the conversion path and a bit of indirection everywhere dates are handled, so Im not sure the dependency savings are worth the added complexity and the small runtime cost. 

On top of that, the newly ported `MapperDateUtils` uses joda to parse and format plain jdk dates, so making joda fully optional there means reimplementing it on java.time and proving it stays byte for byte compatible, which is the real risk. 

If you think this is worth it, I'd argue this refactor should come later in the project, both because it'll be easier to test once we have a PR that compiles, and because its a separate concern.

Comment (dagnir) [reply services-custom/dynamodb-mapper/src/main/java/software/amazon/awssdk/mapper/dynamodb/MapperDateUtils.java:25]: Okay, let's punt this decision; please create a backlog item for revisit this.

**Review by dagnir — APPROVED** [review]
